test/python/circuit/test_parameters.py

```diff
@@ -1783,6 +1781,137 @@ def test_parameter_symbol_equal_after_ufunc(self):
         self.assertEqual(phi._parameter_symbols, cos_phi._parameter_symbols)
 
 
+class TestParameterReferences(QiskitTestCase):
+    """Test the ParameterReferences class."""
+
+    def test_equal_inst_diff_instance(self):
+        """Different value equal instructions are treated as distinct."""
+
+        theta = Parameter("theta")
+        gate1 = RZGate(theta)
+        gate2 = RZGate(theta)
+
+        self.assertIsNot(gate1, gate2)
+        self.assertEqual(gate1, gate2)
+
+        refs = ParameterReferences(((gate1, 0), (gate2, 0)))
+
+        # test __contains__
+        self.assertIn((gate1, 0), refs)
+        self.assertIn((gate2, 0), refs)
+
+        gate_ids = {id(gate1), id(gate2)}
+        self.assertTrue(all(id(gate) in gate_ids for gate, _ in refs))
+        self.assertTrue(all(idx == 0 for _, idx in refs))
+
+    def test_equal_inst_same_instance(self):
+        """Referentially equal instructions are treated as same."""
+
+        theta = Parameter("theta")
+        gate = RZGate(theta)
+
+        refs = ParameterReferences(((gate, 0), (gate, 0)))
+
+        self.assertIn((gate, 0), refs)
+        self.assertEqual(len(refs), 1)
+        self.assertIs(next(iter(refs))[0], gate)
+        self.assertEqual(next(iter(refs))[1], 0)
+
+    def test_extend_refs(self):
+        """Extending references handles duplicates."""
+
+        theta = Parameter("theta")
+        ref0 = (RZGate(theta), 0)
+        ref1 = (RZGate(theta), 0)
+        ref2 = (RZGate(theta), 0)
+
+        refs = ParameterReferences((ref0,))
+
+        refs |= ParameterReferences((ref0, ref1, ref2, ref1, ref0))
+
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
+
+    def test_copy_param_refs(self):
+        """Copy of parameter references is a shallow copy."""
+
+        theta = Parameter("theta")
+        ref0 = (RZGate(theta), 0)
+        ref1 = (RZGate(theta), 0)
+        ref2 = (RZGate(theta), 0)
+        ref3 = (RZGate(theta), 0)
+
+        refs = ParameterReferences((ref0, ref1))
+        refs_copy = refs.copy()
+        refs.add(ref2)
+
+        # check copy not modified
+        self.assertNotIn(ref2, refs_copy)
+        self.assertEqual(list(refs_copy), [ref0, ref1])
+
+        refs_copy.add(ref3)
+
+        # check original not modified
+        self.assertNotIn(ref3, refs)
+        self.assertEqual(list(refs), [ref0, ref1, ref2])
```

Comment:
This test says it's testing for shallow copies, but then it uses logical equality checks rather than referential equality tests when comparing the values stored.  If we fully want to ensure that it's not just a copy (which is currently tested), but specifically a _shallow_ copy, then we probably need the referential equality tests.